test/integration/nodejs/nodejs_test.go

```diff
@@ -0,0 +1,174 @@
+// Copyright 2022-2024 The Parca Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package nodejs
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/relabel"
+	"github.com/stretchr/testify/require"
+	"github.com/testcontainers/testcontainers-go"
+
+	"github.com/parca-dev/parca-agent/pkg/agent"
+	"github.com/parca-dev/parca-agent/pkg/logger"
+	"github.com/parca-dev/parca-agent/pkg/objectfile"
+	"github.com/parca-dev/parca-agent/pkg/profiler/cpu"
+	"github.com/parca-dev/parca-agent/test/integration"
+)
+
+func TestNodeJS(t *testing.T) {
+	ok, _, err := agent.PreflightChecks(false, false, false)
+	require.Truef(t, ok, "preflight checks failed: %v", err)
+	if err != nil {
+		t.Logf("preflight checks passed but with errors: %v", err)
+	}
+
+	tests := []struct {
+		images  map[string]string
+		program string
+		want    []string
+		wantErr bool
+	}{
+		{
+			images: map[string]string{
+				"17": "17-alpine",
+			},
+			program: "testdata/cpu_hog.js",
+			want:    []string{"<main>", "a", "b", "c", "d", "<native code>"},
+			wantErr: false,
+		},
+	}
+	for _, tt := range tests {
+		for version, imageTag := range tt.images {
+			var (
+				program = tt.program
+				want    = tt.want
+				name    = fmt.Sprintf("%s on nodejs-%s", imageTag, version)
+				version = version
+			)
+			t.Run(name, func(t *testing.T) {
+				// Start a Node.js container.
+				ctx, cancel := context.WithCancel(context.Background())
+				t.Cleanup(cancel)
+
+				nodejs, err := testcontainers.GenericContainer(ctx, testcontainers.GenericContainerRequest{
+					ContainerRequest: testcontainers.ContainerRequest{
+						Image: fmt.Sprintf("node:%s", imageTag),
+						Files: []testcontainers.ContainerFile{
+							{
+								HostFilePath:      program,
+								ContainerFilePath: "/cpu_hug.js",
+								FileMode:          0o700,
+							},
+						},
+						Cmd: []string{"node", "/cpu_hug.js"},
```

Comment:
And we need special flags until we implement a dedicated unwinder.

This should help you https://www.polarsignals.com/docs/nodejs

Reply:
I added  --perf-basic-prof-only-functions --interpreted-frames-native-stack for node 21.

But I always got no samples assert failures in test,  the logs related is 
```
    nodejs_test.go:104: node.js (21) is running with pid 1072377
level=debug name=parca-agent-tests ts=2024-03-06T23:20:25.445705981Z caller=cpu.go:607 msg="prefetching process info" pid=1072377
level=debug name=parca-agent-tests ts=2024-03-06T23:20:25.44796707Z caller=info.go:294 msg="failed to fetch runtime unwinder information" err="failed to determine runtime unwinder type" pid=1072377
level=debug name=parca-agent-tests ts=2024-03-06T23:20:25.678367194Z caller=cpu.go:612 msg="adding unwind tables" pid=1072377
level=debug name=parca-agent-tests ts=2024-03-06T23:20:25.828825048Z caller=maps.go:1836 component=bpf_maps msg="jit section" pid=1072377
level=debug name=parca-agent-tests ts=2024-03-06T23:20:25.828838274Z caller=maps.go:1836 component=bpf_maps msg="jit section" pid=1072377
level=debug name=parca-agent-tests ts=2024-03-06T23:20:25.846295992Z caller=maps.go:1840 component=bpf_maps msg="special section" pid=1072377
level=debug name=parca-agent-tests ts=2024-03-06T23:20:25.846310692Z caller=maps.go:1840 component=bpf_maps msg="special section" pid=1072377
level=debug name=parca-agent-tests ts=2024-03-06T23:20:25.846373143Z caller=cpu.go:607 msg="prefetching process info" pid=1072377
level=debug name=parca-agent-tests ts=2024-03-06T23:20:25.846394469Z caller=cpu.go:612 msg="adding unwind tables" pid=1072377
level=debug name=parca-agent-tests ts=2024-03-06T23:20:25.846402531Z caller=maps.go:1499 component=bpf_maps msg="process already cached" pid=1072377
level=debug name=parca-agent-tests ts=2024-03-06T23:20:25.846413131Z caller=cpu.go:607 msg="prefetching process info" pid=1072377
level=debug name=parca-agent-tests ts=2024-03-06T23:20:25.846460398Z caller=cpu.go:612 msg="adding unwind tables" pid=1072377
level=debug name=parca-agent-tests ts=2024-03-06T23:20:25.846470193Z caller=maps.go:1499 component=bpf_maps msg="process already cached" pid=1072377
level=debug name=parca-agent-tests ts=2024-03-06T23:20:25.846480963Z caller=cpu.go:607 msg="prefetching process info" pid=1072377
level=debug name=parca-agent-tests ts=2024-03-06T23:20:25.846489656Z caller=cpu.go:612 msg="adding unwind tables" pid=1072377
level=debug name=parca-agent-tests ts=2024-03-06T23:20:25.846496422Z caller=maps.go:1499 component=bpf_maps msg="process already cached" pid=1072377
level=debug name=parca-agent-tests ts=2024-03-06T23:20:25.846506185Z caller=cpu.go:607 msg="prefetching process info" pid=1072377
level=debug name=parca-agent-tests ts=2024-03-06T23:20:25.846514015Z caller=cpu.go:612 msg="adding unwind tables" pid=1072377
level=debug name=parca-agent-tests ts=2024-03-06T23:20:25.846520502Z caller=maps.go:1499 component=bpf_maps msg="process already cached" pid=1072377
```

Could you please review and see where it might go wrong?

Reply:
The mapping was generated, but under PID 1
```
root@server112102 /h/j/parca-agent (nodejs_integration)# docker exec  -it 20ee666ebf75  sh
# ls /tmp
perf-1.map  v8-compile-cache-0
```

Reply:
bpf verbose shows it's always requesting unwind  info
```
bpf_trace_printk: native: [debug] requesting unwind info for PID: 1279729, comm: docker-entrypoi ctx IP: ffffffff84a5a89b
```

Reply:
The aggregatedStack I got now  was  [[JS:~ JS:~executeUserEntryPoint JS:~Module._load JS:~Module.load JS:~Module._extensions..js JS:~Module._compile JS:*]]. Not complete.

@kakkoyun should I assert on this?

Reply:
Do you see this stack after you added those flags?

Reply:
no. this is the output after I added  --perf-basic-prof-only-functions --interpreted-frames-native-stack

Reply:
@kakkoyun  Please  check in https://github.com/parca-dev/parca-agent/actions/runs/8201453202/job/22430337815?pr=2598

Reply:
This is because node js lazyCompiling and make the loop anonymous, let me see how can I adjust the test..

Reply:
Hmm, I suspect nodejs is inlining things. Could we disable inlining?

Reply:
OK. let me test this without inlining